numpy/core/src/umath/ufunc_object.c

```diff
@@ -2089,6 +2109,145 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
         goto fail;
     }
 
+    if (axis) {
```

Comment:
Can this entire thing go into a static helper function?

As a guess, something like `_parse_axis_arg(PyObject* axis, int do_remap[NPY_MAXARGS], int remap_axis[NPY_MAXARGS][NPY_MAXARGS])`?